### PR TITLE
add the gups benchmark

### PR DESCRIPTION
**Pull request by hppritcha**
Add the gups benchmark used for the Open SHMEM 2016
workshop.  This version of the benchmark does the
right GUPS thing, using shmem_long_g/shmem_long_p
to do "atomic" updates to the global table.

A verification pass uses shmem_set_lock/shmem_clear_lock.
The verification pass can be very slow on some systems.

@kseager 
@jdinan 

This needs a bit more work.  Its not working with the make check
test script for some reason, but I'd like to get this up for review now.

Signed-off-by: Howard Pritchard <howardp@lanl.gov>

### CODE REVIEW

File: test/apps/Makefile.am [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+AM_LDFLAGS = $(LIBTOOL_WRAPPER_LDFLAGS)
+
+check_PROGRAMS = $(TESTS)
+
+if ENABLE_LENGTHY_TESTS
+TESTS = gups
+else
+check_PROGRAMS += gups
+endif
+
```
Comment (kseager) [inline test/apps/Makefile.am]: please try adding: (this enabled make check to work for me) 

NPROCS ?= 2
LOG_COMPILER = $(TEST_RUNNER)

**Comment by kseager** [comment]
other than my one comment this looks good to me! It does produce a prototype compiler warning that might be nice to take care of before pulling in. 

thanks for the add!

**Comment by jdinan** [comment]
Some warnings to clean up:

```
gups.c:219:11: warning: comparison of unsigned expression < 0 is always false [-Wtautological-compare]
        while (n < 0)
               ~ ^ ~
gups.c:408:14: warning: format specifies type 'long' but the argument has type 'uint64_t' (aka 'unsigned long long') [-Wformat]
             logTableSize, TableSize );
             ^~~~~~~~~~~~
gups.c:408:28: warning: format specifies type 'long' but the argument has type 'uint64_t' (aka 'unsigned long long') [-Wformat]
             logTableSize, TableSize );
                           ^~~~~~~~~
gups.c:410:14: warning: format specifies type 'long' but the argument has type 'uint64_t' (aka 'unsigned long long') [-Wformat]
             logTableSize, NumProcs, LocalTableSize);
             ^~~~~~~~~~~~
gups.c:410:38: warning: format specifies type 'long' but the argument has type 'int64_t' (aka 'long long') [-Wformat]
             logTableSize, NumProcs, LocalTableSize);
                                     ^~~~~~~~~~~~~~
gups.c:412:81: warning: format specifies type 'long' but the argument has type 'uint64_t' (aka 'unsigned long long') [-Wformat]
    fprintf( outFile, "Default number of updates (RECOMMENDED) = " FSTR64 "\n", NumUpdates_Default);
                                                                                ^~~~~~~~~~~~~~~~~~
gups.c:484:14: warning: format specifies type 'long' but the argument has type 'int64_t' (aka 'long long') [-Wformat]
             GlbNumErrors, TableSize, (GlbNumErrors <= 0.01*TableSize) ?
             ^~~~~~~~~~~~
gups.c:484:28: warning: format specifies type 'long' but the argument has type 'uint64_t' (aka 'unsigned long long') [-Wformat]
             GlbNumErrors, TableSize, (GlbNumErrors <= 0.01*TableSize) ?
                           ^~~~~~~~~
gups.c:295:1: warning: no previous prototype for function 'SHMEMRandomAccess' [-Wmissing-prototypes]
SHMEMRandomAccess(void)
```

Line 219 looks like a possible bug.  The body of that while loop is unreachable because ```n``` is unsigned.  The printf format warnings can be fixed using this approach (http://stackoverflow.com/questions/9225567/how-to-print-a-int64-t-type-in-c).

**Comment by hppritcha** [comment]
I cleaned up some warnings - @jdinan what CFLAGS and/or compiler are you using?

**Comment by jdinan** [comment]
Clang on my Mac with the --enable-picky SOS configure option.

**Comment by jdinan** [comment]
Ugh, recent changes to Portals have broken our build..  :/

**Comment by jdinan** [comment]
Build fixed, Travis clears, merging..

**Pull request merged**